src/NewRelic.OpenTelemetry/NewRelicTraceExporter.cs

```diff
@@ -115,17 +107,75 @@ public override ExportResult Export(in Batch<Activity> batch)
             };
         }
 
-        private List<NewRelicSpan> ToNewRelicSpans(in Batch<Activity> otSpans)
+        private IEnumerable<NewRelicSpanBatch> ToNewRelicSpanBatches(in Batch<Activity> activityBatch)
         {
-            var nrSpans = new List<NewRelicSpan>();
-            var spanIdsToFilter = new List<string>();
+            var spansByResource = GroupByResource(activityBatch);
+            var spanBatches = new List<NewRelicSpanBatch>(spansByResource.Count);
+
+            foreach (var resource in spansByResource)
+            {
+                string? serviceName = null;
+                string? serviceNamespace = null;
+                Dictionary<string, object>? commonProperties = new Dictionary<string, object>();
+
+                commonProperties.Add(NewRelicConsts.AttribNameCollectorName, "newrelic-opentelemetry-exporter");
+                commonProperties.Add(NewRelicConsts.AttribNameInstrumentationProvider, "opentelemetry");
+
+                foreach (var label in resource.Key.Attributes)
+                {
+                    switch (label.Key)
+                    {
+                        case Resource.ServiceNameKey:
+                            serviceName = label.Value as string;
+                            continue;
+                        case Resource.ServiceNamespaceKey:
+                            serviceNamespace = label.Value as string;
+                            continue;
+                    }
+
+                    commonProperties[label.Key] = label.Value;
+                }
+
+                if (!string.IsNullOrWhiteSpace(serviceName))
+                {
+                    serviceName = serviceNamespace != null
+                        ? serviceNamespace + "." + serviceName
+                        : serviceName;
+                }
+                else
+                {
+                    serviceName = _config.ServiceName;
+                }
+
+                if (!string.IsNullOrWhiteSpace(serviceName))
+                {
+                    commonProperties.Add(NewRelicConsts.Tracing.AttribNameServiceName, serviceName!);
+                }
 
-            foreach (var otSpan in otSpans)
+                var spanBatchCommonProperties = new NewRelicSpanBatchCommonProperties(null, commonProperties);
```

Comment:
Is it possible to work with the NewRelicSpanBatchCommonProperties directly and not need the additional commonProperties dictionary?

Reply:
At the moment no. When NewRelicSpanBatchCommonProperties is constructed it just holds on to a reference to the dictionary that it's passed. So it's not really an additional dictionary. However, this API is certainly a bit wonky. I might consider a pass in a separate PR to make this better.